### ISSUE TITLE
Merge simple pull requests with "Rebase and merge"?

### ISSUE THREAD

**Issue by vinriviere** [freemint/freemint#26]
I have noticed that the freemint history has become a mess:
git log --oneline --decorate --graph --all

This seems to be due to the nature of pull requests, which looks like several remote branches merged into master. This looks overkill in simple cases.

For small contributions, what about using the third option if the "Merge pull request" button: "Rebase and merge"? As far as I understand, that would create a new commit on top of master with the contribution, just like manually integrating a patch. This way the master history would stay linear.

Of course, we could still merge pull requests using "Create a merge commit" for more complex contributions with multiple commits, etc.

**Comment by mikrosk** [comment]
We were discussing the same question in another project -- we've agreed on the rule that the author will specify the type of merge he'd like to perform -- squash, merge or rebase. For instance my issue branches are often mess and it makes sense to squash them (not the case for freemint so far, fortunately :)).

Rebase is not always good because it makes harder to revert specific PR.

**Comment by vinriviere** [comment]
@atic-atac Thanks for the recent merges. BTW, have you any opinion about this suggestion? The recent history, after merges, is very hard to follow. On the other hand, if you had used "Rebase and merge" for the last very simple commits, the history would be very short and readable, for the same result.

![history](https://user-images.githubusercontent.com/4436321/27075623-fed2d48c-502a-11e7-8fa2-fb1b6821e34f.png)

**Comment by atic-atac** [comment]
I think it boils down to personal preference. Personally I don't see a problem with merge.

**Comment by mikrosk** [comment]
I'm with @vinriviere however I'd still suggest to let the author to decide -- for instance, many of my PRs are not very clean, i.e. rebase would produce a lot of dumb commits without any real value so for me squash would work best.

**Comment by vinriviere** [comment]
The difference is only in semantics. A Rebase could be seen as an isolated contribution, while a Merge really reflects the reality of the contributor having forked the repository at some time the asked to merge its changes. The visible consequence is the linearity (or not) of the history.

Merge (vs. Rebase) is a perfectly valid strategy. As long as we agree that the above picture is really what we wish, that's fine for me. Feel free to close this issue, it was just a means of writing those facts somewhere.

**Comment by vinriviere** [comment]
@mikrosk Except for very special tasks, I usually provide very small and specific patches. So my preference for my own commits would be Rebase, as my actual fork/merge history is useless. I agree, it may be different for contributions of other people. Unless specified, I think it should be up to the maintainer to chose the more appropriate way to integrate the contribution. And thanks to GitHub, it is really easy to chose one option or the other.

**Comment by atic-atac** [comment]
If the author of the patch can choose the default merge operation that would be beneficial. Letting the maintainer choose based on authors preference is a headache for the maintainer trying to remember everyones preferences.

**Comment by mikrosk** [comment]
@atic-atac @vinriviere I meant it literally: "hey, this is my PR for fixing XY, please merge as squash/rebase/merge commit". So the maintainer doesn't need to remember anything.

**Comment by mikrosk** [comment]
Btw I have discovered an option which makes it a bit clearer: `--first-parent`. So @vinriviere's example 
becomes `git log --first-parent --decorate --graph --all` and with `--oneline` you'll get a linear (yet incomplete) log:

![screenshot_20171030_205408](https://user-images.githubusercontent.com/3062260/32164760-92580b9a-bdb4-11e7-84e5-6a1961729aad.png)

However personally I'd still prefer to have just a complete CVS-like log with rebase/squash merges.

**Comment by mikrosk** [comment]
It seems to be settled down, most of the changes are rebased/squashed, some (longer) feature branches are merged.
